### DISCUSSION TITLE
Is there a way too -> Set form data & upload bytes at same time?

### DISCUSSION THREAD

**Discussion by andrew-caines** [algesten/ureq#636] [Q&A]
Hello,

I am looking to upload a file to a 3rd party API that requires some form-data be included with the files (bytes).
For example, this is a working version of the requirements inside POSTMAN.

![image](https://github.com/algesten/ureq/assets/12756948/0b7e43fc-d166-47f3-9212-f229408705c7)

I have tried using .send_bytes and it kinda works in the sense that the API will throw and error saying it got the file, but lacked its name and parent folder id.  
This is what I am trying inside rust:
```
//POST https://upload.box.com/api/2.0/files/content
    let file_to_upload = std::fs::read("test.mp3").unwrap(); //TODO make this check to see if file is acutally there/loadable
    let result = agent
        .post("https://upload.box.com/api/2.0/files/content")
        .set("Authorization", &format!("Bearer {}", client.access_token))
        .set("Content-Type", " multipart/form-data")
        .send_form(&[("name", "test.mp3"), ("parent", "0"), ("file", file_to_upload)]);
```

It complains that my file_to_upload is a Vec of bytes and not a string.  Is there a way to mix  .send_form & send_bytes, I honestly tried finding it in the Reference but I couldn't.
Please advise, how would one do this?  (I honestly assume this is a common occurrence/requirement)

Thank you,
Andrew

**Comment by jsha** [comment]
I'm not sure I fully understand the API but it sounds like it wants a
multipart request, where one part contains the JSON attributes and the
second part contains the file contents. Ureq doesn't offer that natively,
but maybe the multipart crate will help?
https://docs.rs/multipart/latest/multipart/

Also I'll note that in your example you are just sending the string
"file.mp3". You'll need to actually read the contents of the file from
disl, then properly encode those bytes amd send them as part of the request.

**Comment by robertosw** [comment]
Why cant we send a file manually? 
1. loading a file as bytes
2. base64 encoding it
3. sending it in "files" in the body of a POST request: `data:image/png;base64,<encoded-bytes-here>`

why is this not possible?

Edit: I've since asked on the rust discord for alternatives where this is possible and learned that the `files` field is not standardized. That would explain why its not possible to send something to exactly this field.

However this only changes my question to: Why cant we send arbitrary data to and field we like? Some endpoints expect some data in such a field

**Reply by algesten** [reply]
ureq can send whatever bytes you want it to send. Hence I don't think this question is not so much about how ureq handles sending files, but how HTTP file sending works in general.

**Reply by robertosw** [reply]
But ureqs send_bytes() always sends them in one specific field (can't remember which, probably "data") 
If we could get full manual control and specify the field name this would be enough I think. 

While searching for it a bit I couldn't find where in ureqs source code this field was specified

**Reply by algesten** [reply]
Not sure what you're thinking about there. When you use `send_bytes()`, it just sends the body bytes. The most it will do is add a header with the length. Example: if you `.send_bytes(b"abc")`, it would look like this on the wire:

```text
POST /mypath HTTP/1.1
Host: myhost.com:80
User-Agent: ureq/2.9.7
Accept: */*
accept-encoding: gzip
Content-Length: 3
abc
```

There is no "data" field, since this is not encoded like a form submission. It's just whatever bytes you want to send.

**Reply by robertosw** [reply]
Oh sorry, I was using www.httbin.org which sends you back your request, but this apparently modifies it a little bit. 
I'll try building the body myself then, thanks :D
